docs/ProofOptimization/ProofOptimization.md

```diff
@@ -0,0 +1,599 @@
+{
+  forall (x: int, y: int, z: int | x == y)
+    ensures x * z == y * z
+  {
+    LemmaMulEquality(x, y, z);
+  }
+}
+```
+
+There was another instance of code in the repository where a proof began to fail after upgrading the version of Z3 included with Dafny. That code originally used `LemmaMulEqualityAuto`. We [updated it](https://github.com/dafny-lang/libraries/blob/7c386fa0b4a267715f9bd49948d1af68e1631e6b/src/dafny/Collections/LittleEndianNat.dfy#L532) to call `LemmaMulEquality` with the appropriate parameters and the proof reliably succeeded.
+
+In general, creating and using explicit lemmas like `LemmaMulEquality`, and skipping lemmas like `LemmaMulEqualityAuto` altogether, tends to lead to proofs with less variability. 
+
+## Adding triggers
+
+If you can't avoid the use of quantifiers, careful use of _triggers_ can help reduce the amount of work Dafny needs to do to benefit from the information in a quantified formula. The original version of the `LemmaMulEqualityAuto` lemma above looks like the following.
+
+<!-- %check-verify -->
+```dafny
+lemma LemmaMulEquality(x: int, y: int, z: int)
+  requires x == y
+  ensures x * z == y * z
+{}
+
+lemma LemmaMulEqualityAuto()
+  ensures forall x: int, y: int, z: int {:trigger x * z, y * z } :: x == y ==> x * z == y * z
+{
+  forall (x: int, y: int, z: int | x == y)
+    ensures x * z == y * z
+  {
+    LemmaMulEquality(x, y, z);
+  }
+}
+```
+
+The [`{:trigger}`](../DafnyRef/DafnyRef#sec-trigger) annotation tells it to only instantiate the formula in the postcondition when the solver sees two existing formulas, `x * z` and `y * z`. That is, if it sees two multiplications in which the right-hand sides are the same.
+
+## Smaller, more isolated definitions
+
+As a general rule, designing a program from the start so that it can be broken into small pieces with clearly-defined interfaces is helpful for human understanding, flexibility, maintenance, and ease of verification. More specific instances of this include the following.
+
+* Break large functions into smaller functions, especially when they're ghost functions, and make at least some of them opaque. I often consider a ghost function of more than a handful of lines to be a code smell. There are exceptions where larger functions are necessary, or the most straightforward alternative, though, such when handling all of the many cases of a complex algebraic data type.
+
+* Prove lemmas about functions rather than having numerous or complex postconditions. The individual lemmas can then be invoked only when needed, and other facts that are true but irrelevant will not bog down the solver.
+
+* Break methods up into smaller pieces, as well. Because methods are always opaque, this will, in many cases, require adding a contract to each broken-out method. To make this more tractable, consider abstracting the concepts used in those contracts into separate functions, some of which may be opaque. 
```

Comment:
I think such a detailed and multi-faceted document deserves a concluding section, especially just to list all the individual techniques for quick reference.